src/sidebar/components/test/tag-editor-test.js

```diff
@@ -1,63 +1,257 @@
+
+      // simulate `keyup` to populate suggestions list
+      wrapper.find('input').simulate('keyup', { key: 'any' });
+      assert.equal(wrapper.find('datalist option').length, 2);
+
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      assert.isFalse(wrapper.find('datalist').exists());
+    });
+
+    it('should not add a tag if the input is empty', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = '';
+
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      addTagsFail();
+    });
+
+    it('should not add a tag if the input is blank space', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = '  ';
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      addTagsFail();
+    });
+
+    it('should not add a tag if its a duplicate of one already in the list', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag1';
+      wrapper.find('input').simulate('keyup'); // simulates a selection
+      addTagsFail();
+    });
 
-    beforeEach(function() {
-      onEditTags = sinon.stub();
-      element = util.createDirective(document, 'tag-editor', {
-        onEditTags: { args: ['tags'], callback: onEditTags },
-        tags: ['foo'],
-      });
-      element.ctrl.onTagsChanged();
+    it('adds a tag via keypress `Enter`', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag3';
+
+      wrapper.find('input').simulate('keypress', { key: 'Enter' });
+      addTagsSuccess(wrapper, ['tag1', 'tag2', 'tag3']);
     });
 
-    it('calls onEditTags handler', function() {
-      assert.calledWith(onEditTags, sinon.match(['foo']));
+    it('adds a tag via keypress `,`', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag3';
+
+      wrapper.find('input').simulate('keypress', { key: ',' });
+      addTagsSuccess(wrapper, ['tag1', 'tag2', 'tag3']);
     });
 
-    it('saves tags to the store', function() {
-      assert.calledWith(fakeTags.store, sinon.match([{ text: 'foo' }]));
+    it('does not add a tag when key is not `,` or  `Enter`', () => {
+      const wrapper = createComponent();
+      wrapper.find('input').instance().value = 'tag3';
+
+      wrapper.find('input').simulate('keypress', { key: 'e' });
+      addTagsFail();
     });
   });
 
-  describe('#autocomplete', function() {
-    it('suggests tags using the `tags` service', function() {
-      const element = util.createDirective(document, 'tag-editor', {
-        tags: [],
-      });
-      element.ctrl.autocomplete('query');
-      assert.calledWith(fakeTags.filter, 'query');
+  context('when removing tags', () => {
+    /**
+     * Helper function to assert that a tag was correctly added
+     */
+    const removeTagsSuccess = tagList => {
```

Comment:
This helper function is only used once, so for ease of human test scanning, it might be better just unrolled into the test.